### DISCUSSION TITLE
General Single Crystal Inelastic Material

### DISCUSSION THREAD

**Discussion by thallor1** [mctools/ncrystal#87] [Q&A] [answered]
Hello - I was wondering if a feature exists within Ncrystal to simulate scattering for a general four dimensional (Qx, Qy, Qz, Energy) space, assuming that the cross section has already been calculated for some grid of points within this space. The ultimate goal is to use this in McStas, but the current implementation (See [here](http://www.mcstas.org/download/components/2.7.1/contrib/Single_crystal_inelastic.html)) is non functional. However, the described functionality is exactly what I'm looking for and I was wondering if this is possible using Ncrystal.

**Comment by tkittel** [comment, marked as answer]
Hi Thomas,

The short answer is unfortunately that, no, NCrystal does not currently support such 4D kernels for single crystal inelastic scattering. While we are certainly aware of the possibility, we simply have not had the manpower available to start looking into that at this point. We are of course always happy to get contributions, although I imagine that this particular issue might have non-trivial challenges in terms of both code, data representation, and input data availability.

I cc @xxcai1 who might have additional comments, and @willend who might know more about the current status of the `Single_crystal_inelastic` component in McStas.

Cheers,
Thomas

**Reply by xxcai1** [reply]
Hi Thomas, Hi Thomas,

It has been an item on my long to-do list for a long time. I think the main problem is that it is difficult to make the data complete. The table may not include enough Brillouin Zones to cover all the possible scattering phase space.  That can introduce "cleaner" signals on detectors. Also due to the cross section being incomplete, the neutron free mean path is likely longer than it should. It may be possible to approximate all the Brillouin Zones by the first one in order to reduce the data size, but this boils down to the manpower problem again. ;-) 

Regards,
Xiao-Xiao

**Comment by thallor1** [comment]
Thanks a lot for the fast replies from both of you. I sort of figured it had not been implemented yet but figured that I should check anyway – this is understandably a difficult task so I will keep an eye on the project and wait. The goal would basically be to simulate complex backgrounds from effects like multiple scattering with true inelastic signals (i.e. phonons / spin waves). Thanks again!
Tom

**Comment by willend** [comment]
Hi @thallor1, 

In regard to Single_crystal_inelastic I am fully aware that it is not perfect, but it was included in the hope of eventually receiving contributions, optimisations etc.

How exactly do you find it "non functional"? We would surely appreciate any direct bug report over at [the McCode issue tracker](https://github.com/McStasMcXtrace/McCode/issues).

I would say that it most certainly works with the original 1.x releases, "should" work with 2.x but not with 3.x.

(And btw "Some Guy" listed in the component text is in fact Duc Le ( @mducle  / https://github.com/mducle ), currently at ISIS in the UK. Perhaps he has tips? :-) )

**Reply by thallor1** [reply]
Hi @willend,

I spoke to Duc about this a while ago and he is the one that said it's a buggy component, I was also trying to use it outside of the suggested use case. I was trying to simulate a delta function in HKL-E space and saw no intensity where I expected to, so I ended up dropping it. It's generally unclear to me what the requirements are for the sqw4 file, I made the one myself with the only line being the HKL-E value that I wanted to simulate the delta function at, but perhaps some interpolation goes on that I interrupted by trying this? 

Perhaps if I tried generating the sqw4 file using iFit as the documentation suggests I would have better luck- I will try this and follow up. If I still don't have luck I will submit to the issue tracker. Thanks!

**Comment by willend** [comment]
Hi @thallor1,

OK, I see. 

For Dirac-delta points in \vec{q}, E it can be that you can use / expand what @granrothge has implemented in his [Spot_sample](http://www.mcstas.org/download/components/2.7.1/contrib/Spot_sample.html)? Or take a look at what @farhi implemented in [ILL_IN4](http://www.mcstas.org/download/components/2.7.1/examples/ILL_IN4.html) - 2D of course, but not knowing details about your case it might be useful?

Wrt. the suggested iFit approach we experimented a bit with it and saw output that "seemed about right". I am not 100% sure we ever arrived at a final conclusion. Please use the latest iFit release of v. [2.0.2](http://ifit.mccode.org/Install.html), don't attempt to use the repo directly - heavy reconstruction going on. There is also a [Docker](https://github.com/McStasMcXtrace/iFit-16.04-Docker) (based on a legacy Ubuntu 16.04...) that may shorten the path to getting a functional iFit these days.

**Reply by willend** [reply]
(Anyway, I seem to have almost made a McStas-oriented ecoup d'etat of this thread, sorry about that! :-) If need be, let us continue the discussion in a dedicated McStas issue in the [McCode issue tracker](https://github.com/McStasMcXtrace/McCode/issues).)

**Comment by tkittel** [comment]
No worries @willend, it is kind of a cross-over issue anyway. But please add a comment with a link here in case the discussion continues in another project, so interested ppl can track what's going on :-)

**Reply by willend** [reply]
This is the place then: https://github.com/McStasMcXtrace/McCode/issues/1251
